% RNA-seq Data Analysis Bootcamp
% Stephen D. Turner, Ph.D.  
  Bioinformatics Core Director
% November 10, 2014

\chapter*{Preface}

This book contains material from a workshop directed toward life scientists with little to no experience with statistical computing or bioinformatics. This will introduce both the Linux/UNIX operating system and the R statistical computing environment, with a focus on a biological application - analyzing RNA-seq data for differentially expressed genes. The first half will introduce basic operation in a UNIX environment, and will cover the first steps in an RNA-seq analysis including QC, alignment, and quantitation. The second half will introduce the R statistical computing environment, and will cover differential gene expression analysis using Bioconductor. At the end of the course or after reading through this book, you will:

1. Know how to provision your own computing resources using Amazon Web Services Elastic Compute Cloud.
2. Be familiar with the UNIX shell, including nagivating the filesystem, creating/examining/removing files, getting help, and batch operations.
3. Know how to align and quantitate gene expression with RNA-seq data.
4. Become familiar with the R statistical computing environment, including data types, variables, array manipulation, functions, data frames, data import/export, visualization, and using packages.
5. Know what packages to use and what steps to take to analyze RNA-seq data for differentially expressed genes.

This book is a PDF version of the online materials available at [bioconnector.github.io](http://bioconnector.github.io/workshops/). Please see the online materials for instructions regarding how to create an Amazon Web Services account and setting up software that will be used in this course. Be sure to stop or terminate any running AWS instances when no longer in use.

This course is sponsored by the Claude Moore Health Sciences Library, and borrows some materials from the Software Carpentry and Data Carpentry projects.
